From eddb1a41b42ee52339d03824a75a08319ef24ad9 Mon Sep 17 00:00:00 2001
From: Mikhail Kshevetskiy <mikhail.kshevetskiy@iopsys.eu>
Date: Wed, 15 Oct 2025 19:01:08 +0300
Subject: [PATCH] net: airoha: use mt7531 mdio for GDM1

Current code just bind mt7531 mdio with it's driver, so mdio device may
not be probed and hense not usable.

This patch
 * force probe mt7531 mdio for GDM1 port
 * set mdio bus for the GDM port (if corresponding mdio/phydev is available)

Signed-off-by: Mikhail Kshevetskiy <mikhail.kshevetskiy@iopsys.eu>
---
 drivers/net/airoha_eth.c | 21 +++++++++++++++++++--
 1 file changed, 19 insertions(+), 2 deletions(-)

diff --git a/drivers/net/airoha_eth.c b/drivers/net/airoha_eth.c
index 29b3a1f0800..dae285a1abc 100644
--- a/drivers/net/airoha_eth.c
+++ b/drivers/net/airoha_eth.c
@@ -12,6 +12,7 @@
 #include <dm/device-internal.h>
 #include <dm/devres.h>
 #include <dm/lists.h>
+#include <eth_phy.h>
 #include <mapmem.h>
 #include <miiphy.h>
 #include <net.h>
@@ -358,6 +359,7 @@ struct airoha_eth {
 
 	struct airoha_eth_soc_data *soc;
 
+	struct udevice *switch_mdio;
 	struct airoha_qdma qdma[AIROHA_MAX_NUM_QDMA];
 };
 
@@ -1013,6 +1015,8 @@ static int airoha_eth_port_probe(struct udevice *dev)
 {
 	struct airoha_eth *eth = (void *)dev_get_driver_data(dev);
 	struct airoha_gdm_port *port = dev_get_priv(dev);
+	struct mdio_perdev_priv *pdata;
+	struct mii_dev *mdio_bus;
 	int ret;
 
 	port->qdma = &eth->qdma[0];
@@ -1021,14 +1025,27 @@ static int airoha_eth_port_probe(struct udevice *dev)
 	if (ret)
 		return ret;
 
+	mdio_bus = NULL;
 	if (port->id > 1) {
 		ret = airoha_pcs_init(dev);
 		if (ret)
 			return ret;
 
 		port->phydev = dm_eth_phy_connect(dev);
+		if (port->phydev)
+			mdio_bus = port->phydev->bus;
+	} else {
+		if (eth->switch_mdio && !device_probe(eth->switch_mdio)) {
+			pdata = dev_get_uclass_priv(eth->switch_mdio);
+			mdio_bus = pdata->mii_bus;
+		}
 	}
 
+#ifdef CONFIG_DM_ETH_PHY
+	if (!IS_ERR_OR_NULL(mdio_bus))
+		eth_phy_set_mdio_bus(dev, mdio_bus);
+#endif
+
 	return 0;
 }
 
@@ -1286,8 +1303,8 @@ static ofnode an7583_get_scu_node(struct udevice *dev)
 static int airoha_eth_bind(struct udevice *dev)
 {
 	struct airoha_eth_soc_data *data = (void *)dev_get_driver_data(dev);
+	struct airoha_eth *eth = dev_get_priv(dev);
 	ofnode switch_node, mdio_node;
-	struct udevice *mdio_dev;
 	int ret = 0;
 
 	/*
@@ -1313,7 +1330,7 @@ static int airoha_eth_bind(struct udevice *dev)
 	}
 
 	ret = device_bind_driver_to_node(dev, "mt7531-mdio", "mdio",
-					 mdio_node, &mdio_dev);
+					 mdio_node, &eth->switch_mdio);
 	if (ret)
 		debug("Warning: failed to bind mdio controller\n");
 
-- 
2.51.0

